pkg/auth/oauth/handlers/selectprovider.go

```diff
@@ -0,0 +1,186 @@
+//    "Names.CSRF":      form.Values.CSRF,
+//    "Names.Username":  form.Names.Username,
+//    "Names.Password":  form.Names.Password,
+//    "Values.Then":     form.Values.Then,
+//    "Values.CSRF":     form.Values.CSRF,
+//    "Values.Username": form.Values.Username,
+//  }
+
+//  for field, value := range testFields {
+//    if !bytes.Contains(output, []byte(value)) {
+//      allErrs = append(allErrs, errors.New(fmt.Sprintf("template is missing parameter {{ .%s }}", field)))
+//    }
+//  }
+
+//  return allErrs
+// }
+
+type selectProviderTemplateRenderer struct {
+	selectProviderTemplate *template.Template
+}
+
+func (r selectProviderTemplateRenderer) Render(redirectors map[string]AuthenticationRedirector, w http.ResponseWriter, req *http.Request) {
+	w.Header().Add("Content-Type", "text/html")
+	w.WriteHeader(http.StatusOK)
+	if err := r.selectProviderTemplate.Execute(w, Providers{Redirectors: redirectors}); err != nil {
+		util.HandleError(fmt.Errorf("unable to render select provider template: %v", err))
+	}
+}
+
+// SelectProviderTemplateExample is a basic template for customizing the provider selection page.
+const SelectProviderTemplateExample = `<!DOCTYPE html>
+<!--
+
+This template can be modified and used to customize the provider selection page. To replace
+the provider selection page, set master configuration option oauthConfig.templates.selectProvider to
+the path of the template file. Don't remove parameters in curly braces below.
+
+oauthConfig:
+  templates:
+    selectProvider: templates/select-provider-template.html
+
+-->
+<html>
+  <head>
+    <title>Select a provider</title>
+    <style type="text/css">
+      body {
+        font-family: "Open Sans", Helvetica, Arial, sans-serif;
+        font-size: 14px;
+        margin: 15px;
+      }
+
+      input {
+        margin-bottom: 10px;
+        width: 300px;
+      }
+
+      .error {
+        color: red;
+        margin-bottom: 10px;
+      }
+    </style>
+  </head>
+  <body>
+
+    {{ range $redirectorName, $redirector := .Redirectors }}
+      <div class="error">{{ $redirectorName }}</div>
+    {{ end }}
+
+  </body>
+</html>
+`
+
+var defaultSelectProviderTemplate = template.Must(template.New("defaultSelectProvider").Parse(defaultSelectProviderTemplateString))
+
+const defaultSelectProviderTemplateString = `<!DOCTYPE html>
+<html>
+  <head>
+    <title>Login</title>
+    <style type="text/css">
+      body {
+        font-family: "Open Sans", Helvetica, Arial, sans-serif;
+        font-size: 14px;
+        margin: 15px;
+      }
+    </style>
+  </head>
+  <body>
+
+    {{ range $redirectorName, $redirector := .Redirectors }}
+      <!-- TODO JWF yuck, maybe we shouldn't rely on this happening within the template and pass something in instead  -->
+      <a href="javascript: window.location.search = window.location.search + ((window.location.search.length > 0) ? '&' : '?') + 'useRedirectHandler={{ $redirectorName }}'">{{ $redirectorName }}</div>
```

Comment:
@liggitt thoughts here?  i would be less concerned if it was only our own templates, i would worry about a custom template screwing this up